metrics/registry.go

```diff
@@ -113,45 +119,67 @@ func (registry *registryImpl) FuncGauge(name string, f func() int64) Gauge {
 }
 
 func (registry *registryImpl) Meter(name string) Meter {
-	metric, present := registry.metricMap.Get(name)
-	if present {
-		meter, ok := metric.(Meter)
-		if !ok {
-			panic(fmt.Errorf("metric '%v' already exists and is not a meter. It is a %v", name, reflect.TypeOf(metric).Name()))
+	metric := registry.getRefCounted(name, func() refCounted {
+		return &meterImpl{
+			Meter:    metrics.NewMeter(),
+			registry: registry,
+			name:     name,
 		}
-		return meter
-	}
+	})
 
-	meter := &meterImpl{
-		Meter: metrics.NewMeter(),
-		dispose: func() {
-			registry.dispose(name)
-		},
+	meter, ok := metric.(Meter)
+	if !ok {
+		panic(fmt.Errorf("metric '%v' already exists and is not a meter. It is a %v", name, reflect.TypeOf(metric).Name()))
```

Comment:
do we really want a panic here?

Reply:
It should never happen, but if it does happen, it should happen very quickly, as soon as either the controller starts up or a router starts up or a router connects. We don't have any metrics that would be initialized late. So if it passes smoke, it should be good. It's a fast fail for a programming mistake.